File: src/skill_xray/disposition.py
"""Explicit operator decisions, separate from raw evidence and model opinions."""

from __future__ import annotations

import re
from copy import deepcopy
from dataclasses import asdict
from pathlib import PurePosixPath

from .checks.coverage import _static_severity, is_inventory_note
from .correlate import canonical, source_region
from .findings import SEVERITY_RANK
from .ingest import _BENIGN_LEDGER

POLICY_VERSION = "skill-xray/scoped-policy/v1"
_SCOPE = ("rule_id", "path", "fingerprint", "context_digest")


def _policy_entries(policy):
    if policy is None:
        return {}
    if (not isinstance(policy, dict) or set(policy) != {"version", "decisions"}
            or policy["version"] != POLICY_VERSION or not isinstance(policy["decisions"], list)
            or len(policy["decisions"]) > 512):
        raise ValueError("Invalid operator policy")
    entries = {}
    for entry in policy["decisions"]:
        required = set(_SCOPE) | {"action", "reason"}
        if isinstance(entry, dict) and entry.get("action") == "demote":
            required.add("effective_severity")
        if (not isinstance(entry, dict) or set(entry) != required
                or any(not isinstance(value, str) or not value.strip() for value in entry.values())
                or entry["action"] not in {"suppress", "demote"}
                or len(entry["reason"]) > 1024
                or any(not re.fullmatch(r"[0-9a-f]{64}", entry[key])
                       for key in ("fingerprint", "context_digest"))
                or PurePosixPath(entry["path"]).is_absolute()
                or ".." in entry["path"].split("/")
                or entry.get("effective_severity", "low") not in SEVERITY_RANK):
            raise ValueError("Invalid operator policy decision")
        # Paths are exact IR identities, never filesystem paths to open or normalize.
        key = tuple(entry[field] for field in _SCOPE)
        if key in entries:
            raise ValueError("Duplicate operator policy scope")
        entries[key] = entry
    return entries


def _source_known(result, parsed):
    finding = result["finding"]
    artifact = parsed.by_rel.get(finding["path"])
    if artifact is None or artifact.diagnostics:
        return False
    offset, length = finding.get("offset"), finding.get("length")
    if offset is not None:
        return (artifact.raw is not None and type(offset) is int and offset >= 0
                and type(length) is int and length > 0 and offset + length <= len(artifact.raw))
    line, column = finding.get("line"), finding.get("column")
    lines = artifact.text.split("\n") if artifact.text is not None else []
    if type(line) is not int or not 1 <= line <= len(lines):
        return False
    start = {"line": line, "col": column if column is not None else 1}
    evidence = finding.get("evidence", {})
    end = evidence["end"] if "end" in evidence else start
    try:
        source_region(artifact, start, end, byte_columns=evidence.get("engine") == "opengrep")
    except (ValueError, TypeError):
        return False
    return True


def _material_ledger_entry(entry, parsed):
    if (entry.get("phase") or "static") != "static":
        return True
    reason = str(entry.get("reasonCode") or "unknown")
    artifact = parsed.by_rel.get(str(entry.get("path") or ""))
    return (reason not in _BENIGN_LEDGER
            and _static_severity(reason, artifact.kind if artifact else None) is not None)


def apply_dispositions(parsed, correlated, triads, *, policy=None, context_errors=()):
    """Retain by default. Only exact operator scope may suppress or lower severity.

    All results remain in the audit list, including suppressed results. Unknown capability
    axes are explanation, not permission. Material gaps block policy package-wide:
    an unanalyzed artifact may change cross-file capability or flow context.
    """
    entries = _policy_entries(policy)
    final = deepcopy(correlated)
    raw = final["raw_candidates"]
    contexts = {key: asdict(triad) for key, triad in sorted(triads.items())}
    for context in contexts.values():
        context["evidence"].sort(key=canonical)
    final["capability_contexts"] = contexts
    context_limits = [{"manifest": key, "limitations": sorted(value.limitations)}
                      for key, value in sorted(triads.items()) if value.limitations]
    package_gap = bool(context_errors or context_limits or any(
        _material_ledger_entry(entry, parsed) for entry in parsed.ledger_exceptions) or any(
        (not item["finding"]["vector"] or item.get("coverage") != "no-reported-gap")
        and not is_inventory_note(item["finding"]) for item in raw))
    decisions = {}
    for result in final["results"]:
        finding = result["finding"]
        context = contexts.get(result["manifest"] or "")
        incomplete = (package_gap or result["manifest"] is None
                      or context is None or bool(context["limitations"])
                      or (not is_inventory_note(finding)
                          and (bool(result["limitations"]) or not _source_known(result, parsed))))
        protected = (not finding["vector"] or any(
            item["provenance"] != "deterministic-check-output" for item in result["provenance"]))
        disposition, reason, provenance = (
            "reported", "Original deterministic evidence retained", "deterministic-policy")
        original = effective = finding["severity"]
        entry = entries.get((result["rule_id"], finding["path"], result["fingerprint"],
                             result["context_digest"]))
        if incomplete or protected:
            reason = "Incomplete context or protected operational evidence; retained"
        elif entry is not None:
            if entry["action"] == "suppress":
                disposition = "suppressed"
            elif SEVERITY_RANK[entry["effective_severity"]] > SEVERITY_RANK.get(original, 99):
                disposition, effective = "corrected", entry["effective_severity"]
            if disposition != "reported":
                reason, provenance = entry["reason"], "operator-policy"
            else:
                reason = "Policy cannot raise severity or make a no-op correction; retained"
        result.update(disposition=disposition, decision_reason=reason,
                      decision_provenance=provenance, policy_version=POLICY_VERSION,
                      original_severity=original, effective_severity=effective,
                      capability_context={key: deepcopy(value) for key, value in context.items()
                                          if key != "evidence"} if context is not None else None,
                      coverage="incomplete" if incomplete
                      else "no-reported-gap")
        decisions[result["id"]] = result
    for link in final["links"]:
        result = decisions[link["result_id"]]
        if link["disposition"] != "duplicate":
            link.update(disposition=result["disposition"], reason=result["decision_reason"])
        link.update(policy_version=POLICY_VERSION,
                    provenance="deterministic-correlation" if link["disposition"] == "duplicate"
                    else result["decision_provenance"])
    final["coverage"] = "incomplete" if package_gap or any(
        r["coverage"] == "incomplete" for r in final["results"]) else "no-reported-gap"
    final["context_limitations"] = context_limits
    final["execution_successful"] = not bool(context_errors or any(
        not item["finding"]["vector"] and item["finding"]["severity"] in {"critical", "high"}
        for item in raw))
    return final


LLM_APPLY_VERSION = "skill-xray/llm-apply/v1"
# Only the text-pattern directive vectors the review contracts cover. Code-lane, taint, byte
# forensics and every other mechanically anchored vector are never model-adjustable.
LLM_APPLY_VECTORS = frozenset({"SXV-028", "SXV-029", "SXV-030", "SXV-031"})


def apply_llm_review(final, decisions, *, effective_severity="low"):
    """Opt-in: a validated ``llm-disputed`` review demotes its result to ``effective_severity``.

    Only the one text-pattern result the dispute covers, only for vectors in the review
    contracts, and never a result the deterministic policy already protected (incomplete
    context, non-deterministic provenance, coverage gap). It never suppresses and never raises.
    The result stays in the audit as ``corrected`` with ``llm-review-policy`` provenance, like an
    operator demote.
    """
    if effective_severity not in SEVERITY_RANK:
        raise ValueError("Invalid LLM demotion severity")
    disputed = {d["candidate_id"]: d for d in decisions
                if d.get("disposition") == "llm-disputed" and d.get("status") == "proposed"}
    results = {r["id"]: r for r in final["results"]}
    applied = 0
    for link in final["links"]:
        decision, result = disputed.get(link["candidate_id"]), results.get(link["result_id"])
        if decision is None or result is None or result.get("llm_applied"):
            continue
        finding = result["finding"]
        if (result["disposition"] != "reported" or result["coverage"] != "no-reported-gap"
                or finding["vector"] not in LLM_APPLY_VECTORS
                or any(p["provenance"] != "deterministic-check-output"
                       for p in result["provenance"])
                or SEVERITY_RANK[effective_severity] <= SEVERITY_RANK.get(finding["severity"], 99)):
            continue
        result.update(disposition="corrected", effective_severity=effective_severity,
                      decision_reason=decision.get("reason") or "LLM review disputed the finding",
                      decision_provenance="llm-review-policy", policy_version=LLM_APPLY_VERSION,
                      llm_applied=True, llm_candidate_id=link["candidate_id"])
        applied += 1
    for link in final["links"]:
        result = results[link["result_id"]]
        if link["disposition"] != "duplicate" and result.get("llm_applied"):
            link.update(disposition="corrected", reason=result["decision_reason"],
                        provenance="llm-review-policy", policy_version=LLM_APPLY_VERSION)
    final["llm_applied"] = applied
    return final

File: internal/correlate/disposition.go
package correlate

import (
	"cmp"
	"errors"
	"maps"
	"regexp"
	"slices"
	"strings"
	"unicode/utf8"

	"github.com/traceforce/skill-xray/internal/capability"
	"github.com/traceforce/skill-xray/internal/checks"
	"github.com/traceforce/skill-xray/internal/findings"
	"github.com/traceforce/skill-xray/internal/ingest"
	"github.com/traceforce/skill-xray/internal/parse"
	"github.com/traceforce/skill-xray/internal/pytext"
)

const (
	PolicyVersion   = "skill-xray/scoped-policy/v1"
	LLMApplyVersion = "skill-xray/llm-apply/v1"
)

// LLMApplyVectors are the text-pattern directive vectors the review contracts cover; every
// mechanically anchored vector is never model-adjustable.
var LLMApplyVectors = map[string]bool{"SXV-028": true, "SXV-029": true, "SXV-030": true, "SXV-031": true}

// scope is (rule_id, path, fingerprint, context_digest): paths are exact IR identities.
type scope [4]string

var hexDigest = regexp.MustCompile(`^[0-9a-f]{64}\z`)

// policyEntries validates the whole operator policy before any decision applies.
func policyEntries(policy map[string]any) (map[scope]map[string]string, error) {
	if policy == nil {
		return nil, nil
	}
	decisions, ok := policy["decisions"].([]any)
	if len(policy) != 2 || policy["version"] != any(PolicyVersion) || !ok || len(decisions) > 512 {
		return nil, errors.New("Invalid operator policy")
	}
	entries := map[scope]map[string]string{}
	for _, item := range decisions {
		entry, isMap := item.(map[string]any)
		required := map[string]bool{"rule_id": true, "path": true, "fingerprint": true, "context_digest": true,
			"action": true, "reason": true}
		if isMap && entry["action"] == any("demote") {
			required["effective_severity"] = true
		}
		values, valid := map[string]string{}, isMap && len(entry) == len(required)
		for k, v := range entry {
			s, isStr := v.(string)
			if !required[k] || !isStr || pytext.Strip(s) == "" {
				valid = false
				break
			}
			values[k] = s
		}
		if !valid || values["action"] != "suppress" && values["action"] != "demote" ||
			utf8.RuneCountInString(values["reason"]) > 1024 ||
			!hexDigest.MatchString(values["fingerprint"]) || !hexDigest.MatchString(values["context_digest"]) ||
			strings.HasPrefix(values["path"], "/") || slices.Contains(strings.Split(values["path"], "/"), "..") ||
			findings.Rank(cmp.Or(values["effective_severity"], "low"), -1) < 0 {
			return nil, errors.New("Invalid operator policy decision")
		}
		key := scope{values["rule_id"], values["path"], values["fingerprint"], values["context_digest"]}
		if _, dup := entries[key]; dup {
			return nil, errors.New("Duplicate operator policy scope")
		}
		entries[key] = values
	}
	return entries, nil
}

// sourceKnown is true when the result's position is provable against a cleanly parsed artifact.
func sourceKnown(r *Result, p *parse.Package) bool {
	finding := r.Finding
	a := p.ByRel[str(finding, "path")]
	if a == nil || len(a.Diagnostics) > 0 {
		return false
	}
	if offset := finding["offset"]; offset != nil {
		o, okO := offset.(int)
		l, okL := finding["length"].(int)
		return a.Raw != nil && okO && o >= 0 && okL && l > 0 && o+l <= len(a.Raw)
	}
	var lines []string
	if a.Text != nil {
		lines = strings.Split(*a.Text, "\n")
	}
	line, ok := finding["line"].(int)
	if !ok || line < 1 || line > len(lines) {
		return false
	}
	start := map[string]any{"line": line, "col": 1}
	if column := finding["column"]; column != nil {
		start["col"] = column
	}
	ev, _ := finding["evidence"].(map[string]any)
	end := start
	if v, present := ev["end"]; present {
		end, _ = v.(map[string]any)
	}
	_, _, _, err := SourceRegion(a, start, end, ev["engine"] == any("opengrep"), lines)
	return err == nil
}

// materialLedgerEntry is a coverage gap that may hide cross-file context: anything outside the
// static phase, and any static skip that is neither benign nor an opaque asset.
func materialLedgerEntry(e ingest.LedgerEntry, p *parse.Package) bool {
	if cmp.Or(e.Phase, "static") != "static" {
		return true
	}
	reason, kind := cmp.Or(e.ReasonCode, "unknown"), ""
	if a := p.ByRel[e.Path]; a != nil {
		kind = a.Kind
	}
	return !ingest.BenignLedger[reason] && checks.StaticSeverity(reason, kind, e.Path) != ""
}

// triadMap is asdict(triad) with the evidence sorted by canonical form; a deep copy.
func triadMap(t *capability.Triad) map[string]any {
	evidence, _ := pytext.DeepCopy(t.Evidence).([]map[string]any)
	if evidence == nil {
		evidence = []map[string]any{}
	}
	slices.SortFunc(evidence, func(a, b map[string]any) int {
		return cmp.Compare(pytext.Canonical(a), pytext.Canonical(b))
	})
	var manifest any
	if t.Manifest != nil {
		manifest = *t.Manifest
	}
	return map[string]any{"manifest": manifest, "claimed": maps.Clone(t.Claimed), "declared": maps.Clone(t.Declared),
		"observed": maps.Clone(t.Observed), "evidence": evidence, "limitations": append([]string{}, t.Limitations...)}
}

// clone copies what the disposition phase rewrites (results, links); raw candidates, finding
// documents and evidence are shared, as no phase edits them.
func (c *Correlation) clone() *Correlation {
	out := *c
	out.Results = make([]*Result, len(c.Results))
	for i, r := range c.Results {
		copied := *r
		out.Results[i] = &copied
	}
	out.Links = slices.Clone(c.Links)
	return &out
}

// ApplyDispositions retains every result by default; only an exact operator scope may suppress
// or lower severity, and never while the package context is incomplete or the evidence is
// operational. Every result stays in the audit list. The error is the policy ValueError.
func ApplyDispositions(p *parse.Package, c *Correlation, triads map[string]*capability.Triad, policy map[string]any, contextErrors []string) (*Correlation, error) {
	entries, err := policyEntries(policy)
	if err != nil {
		return nil, err
	}
	final := c.clone()
	contexts, limits := map[string]map[string]any{}, []ContextLimit{}
	for _, key := range slices.Sorted(maps.Keys(triads)) {
		contexts[key] = triadMap(triads[key])
		if t := triads[key]; len(t.Limitations) > 0 {
			limits = append(limits, ContextLimit{key, slices.Sorted(slices.Values(t.Limitations))})
		}
	}
	packageGap := len(contextErrors) > 0 || len(limits) > 0 ||
		slices.ContainsFunc(p.LedgerExceptions, func(e ingest.LedgerEntry) bool { return materialLedgerEntry(e, p) }) ||
		slices.ContainsFunc(final.RawCandidates, func(c Candidate) bool {
			return (str(c.Finding, "vector") == "" || c.Coverage != "no-reported-gap") && !isInventoryNote(c.Finding)
		})
	byID := map[string]*Result{}
	for _, r := range final.Results {
		finding, manifestKey := r.Finding, ""
		if r.Manifest != nil {
			manifestKey = *r.Manifest
		}
		context, hasContext := contexts[manifestKey]
		incomplete := packageGap || r.Manifest == nil || !hasContext || len(context["limitations"].([]string)) > 0 ||
			!isInventoryNote(finding) && (len(r.Limitations) > 0 || !sourceKnown(r, p))
		protected := str(finding, "vector") == "" || slices.ContainsFunc(r.Provenance, func(pr Provenance) bool {
			return pr.Provenance != "deterministic-check-output"
		})
		severity := str(finding, "severity")
		d := &Decision{Disposition: "reported", DecisionReason: "Original deterministic evidence retained",
			DecisionProvenance: "deterministic-policy", PolicyVersion: PolicyVersion,
			OriginalSeverity: severity, EffectiveSeverity: severity, Coverage: "no-reported-gap"}
		entry := entries[scope{r.RuleID, str(finding, "path"), r.Fingerprint, r.ContextDigest}]
		switch {
		case incomplete || protected:
			d.DecisionReason = "Incomplete context or protected operational evidence; retained"
		case entry != nil:
			if entry["action"] == "suppress" {
				d.Disposition = "suppressed"
			} else if findings.SeverityRank[entry["effective_severity"]] > findings.Rank(severity, 99) {
				d.Disposition, d.EffectiveSeverity = "corrected", entry["effective_severity"]
			}
			if d.Disposition != "reported" {
				d.DecisionReason, d.DecisionProvenance = entry["reason"], "operator-policy"
			} else {
				d.DecisionReason = "Policy cannot raise severity or make a no-op correction; retained"
			}
		}
		if hasContext {
			d.CapabilityContext = triadMap(triads[manifestKey])
			delete(d.CapabilityContext, "evidence")
		}
		if incomplete {
			d.Coverage = "incomplete"
		}
		r.Decision = d
		byID[r.ID] = r
	}
	for i := range final.Links {
		link := &final.Links[i]
		r := byID[link.ResultID]
		if link.Disposition != "duplicate" {
			link.Disposition, link.Reason = r.Disposition, r.DecisionReason
		}
		link.LinkDecision = &LinkDecision{PolicyVersion: PolicyVersion, Provenance: r.DecisionProvenance}
		if link.Disposition == "duplicate" {
			link.Provenance = "deterministic-correlation"
		}
	}
	final.Applied = &Applied{CapabilityContexts: contexts, Coverage: "no-reported-gap", ContextLimitations: limits,
		ExecutionSuccessful: len(contextErrors) == 0 && !slices.ContainsFunc(final.RawCandidates, func(c Candidate) bool {
			sev := str(c.Finding, "severity")
			return str(c.Finding, "vector") == "" && (sev == "critical" || sev == "high")
		})}
	if packageGap || slices.ContainsFunc(final.Results, func(r *Result) bool { return r.Coverage == "incomplete" }) {
		final.Coverage = "incomplete"
	}
	return final, nil
}

// Review is what ApplyLLMReview reads of an llm.Decision (candidate_id, disposition, status,
// reason); llm imports this package, so the full type cannot be named here.
type Review struct {
	CandidateID, Disposition, Status, Reason string
}

// ApplyLLMReview demotes to "low" the one text-pattern result a validated "llm-disputed" review
// covers, in place: never a result the deterministic policy protected, never a suppression,
// never a raise. The result stays in the audit as "corrected" with "llm-review-policy".
func ApplyLLMReview(c *Correlation, decisions []Review) *Correlation {
	disputed := map[string]Review{}
	for _, d := range decisions {
		if d.Disposition == "llm-disputed" && d.Status == "proposed" {
			disputed[d.CandidateID] = d
		}
	}
	results := map[string]*Result{}
	for _, r := range c.Results {
		results[r.ID] = r
	}
	applied := 0
	for _, link := range c.Links {
		decision, ok := disputed[link.CandidateID]
		r := results[link.ResultID]
		if !ok || r == nil || r.LLMApplied {
			continue
		}
		if r.Disposition != "reported" || r.Coverage != "no-reported-gap" || !LLMApplyVectors[str(r.Finding, "vector")] ||
			slices.ContainsFunc(r.Provenance, func(p Provenance) bool {
				return p.Provenance != "deterministic-check-output" || p.Analyzer == "opengrep"
			}) || findings.SeverityRank["low"] <= findings.Rank(str(r.Finding, "severity"), 99) {
			continue
		}
		r.Disposition, r.EffectiveSeverity = "corrected", "low"
		r.DecisionReason = cmp.Or(decision.Reason, "LLM review disputed the finding")
		r.DecisionProvenance, r.PolicyVersion = "llm-review-policy", LLMApplyVersion
		r.LLMApplied, r.LLMCandidateID = true, link.CandidateID
		applied++
	}
	for i := range c.Links {
		link := &c.Links[i]
		if r := results[link.ResultID]; link.Disposition != "duplicate" && r.LLMApplied {
			link.Disposition, link.Reason = "corrected", r.DecisionReason
			link.LinkDecision = &LinkDecision{PolicyVersion: LLMApplyVersion, Provenance: "llm-review-policy"}
		}
	}
	c.LLMApplied = &applied
	return c
}
